### PR TITLE
deck: Assigned unique id to each card.

### LINKED ISSUE #4: Make it possible to uniquely identify cards
# Description
The `makeCard` function in `deck.js` should assign a unique id to each card to ensure each card can be uniquely identified.

# Current Code Snippet
```javascript
/**
 * Helper function to make a card object.
 * @param {string} type one of ['number', 'special', 'wild']
 * @param {string} color one of ['red', 'yellow', 'green', 'blue']
 * @param {string} value the numeric value as string or the special card name ("skip", "reverse", "draw2", "wild", "draw4")
 * @returns {{type: string, color: string, value: string, id: unknown}} card - An object representing an UNO card.
 */
function makeCard(type, color, value) {
    //todo: Implement unique identification of cards by assigning an id to each card
    return { type, color, value };
}
```
# Expected Output
Each card object returned by makeCard should include a unique id property.

Some discussion is needed on the type of unique identification used. We can choose from UUID, random numbers etc.
Please propose your approach here and you will be assigned accordingly.

# Notes
* Ensure the id is unique for every card created.
* Update the JSDoc accordingly.

### PR DESCRIPTION
**Pull request by sethdivyansh**
- Created an ID in the format `card-type-color-value`.
- Added a `sameCardCount` array to uniquely identify cards of the same type and color.
- Generated a unique ID using the format `card-type-color-value-sameCardCount[id]`.

Fixes #4 

<h3>Reason for choosing custom unique id instead of UUID etc</h3>

1. Custom unique id contains meaningful information about the card that may help in future debuging. For example: `card-Special-Red-skip-2` immediately conveys that the card is a red "skip" card and it is the second one of its kind.
2. Generating custom id is computationally cheaper and simpler compared to generating a UUID.
3. IDs like `card-Special-Red-skip-2` are `human-readable`, making it easier for developers and testers to trace issues, verify game logic, and understand game states during testing and development. `UUIDs`, being long and non-descriptive, `lack this readability`.

<details>
<summary>Self-review checklist</summary>

- [x] Self reviewed the changes for clarity and maintainability (variable names, code reuse, readability, etc.).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.
- [ ] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

### CODE REVIEW

File: backend/uno-game-engine/deck.js [hunk 1]
```diff
@@ -60,7 +60,16 @@ export default function getShuffledCardDeck() {
  */
 function makeCard(type, color, value) {
     //todo: Implement unique identification of cards by assigning an id to each card
-    return { type, color, value };
+    let card_num = [];
```
[the remark this answers is not in the document]
Comment (sethdivyansh) [reply backend/uno-game-engine/deck.js]: Yes, I realized my mistake later. I have corrected it; please review the updated version.

**Comment by kuv2707** [comment]
@sethdivyansh Please rebase your branch against `upstream/main` and resolve the merge conflicts. 
You'd also have to deal with types now that this module is converted to TS.

I'l merge it then.

**Comment by sethdivyansh** [comment]
@kuv2707 I have resolved all the conflicts. Please review the PR and merge it.

File: backend/uno-game-engine/types.d.ts [hunk 2]
```diff
@@ -20,5 +20,5 @@ type UNOCard = {
     type: CardType;
     color: CardColor;
     value: CardValue;
-    id: undefined;
+    id: string;
```
Comment (kuv2707) [inline backend/uno-game-engine/types.d.ts:23]: The typo should be fixed in the original commit itself. Refer to the docs, where an example is illustrated for this.

Comment (sethdivyansh) [reply backend/uno-game-engine/types.d.ts:23]: done 👍

Comment (kuv2707) [reply backend/uno-game-engine/types.d.ts:23]: Great!
Now the commit message doesn't render well.

A part of the heading extends to the content. Try to write a concise heading.

There are also dome typos.

File: backend/uno-game-engine/deck.ts [hunk 3]
```diff
@@ -57,7 +58,13 @@ export default function getShuffledCardDeck(): Array<UNOCard> {
  */
 function makeCard(type: CardType, color: CardColor, value: CardValue): UNOCard {
     //todo: Implement unique identification of cards by assigning an id to each card
-    return { type, color, value, id: undefined };
+    let id = `card-${type}-${color}-${value}`;
```
Comment (kuv2707) [inline backend/uno-game-engine/deck.ts:61]: Prefer const for variables which do not change. Eslint should have given an error.

File: backend/uno-game-engine/deck.ts [hunk 4]
```diff
@@ -57,7 +58,13 @@ export default function getShuffledCardDeck(): Array<UNOCard> {
...
+
+    if (!sameCardCount[id]) sameCardCount[id] = 0;
+    sameCardCount[id]++; // increment the count of same cards to assign unique id
+
+    let uid = `${id}-${sameCardCount[id]}`;
```
Comment (kuv2707) [inline backend/uno-game-engine/deck.ts:66]: Same here

**Review by kuv2707 — CHANGES_REQUESTED** [review]

**Comment by sethdivyansh** [comment]
@kuv2707 I have made all the changes.

**Comment by kuv2707** [comment]
@sethdivyansh  Please add the "Fixes" clause in the commit message.

**Comment by sethdivyansh** [comment]
Done

**Comment by sethdivyansh** [comment]
> Done

**Comment by kuv2707** [comment]
@sethdivyansh Please add the motivation to use a custom-generated ID instead of approaches like uuid etc in the PR description.
There had been some discussion regarding it in the issue page.

This PR is ready to be merged. Will do it once the event begins officially.

**Comment by sethdivyansh** [comment]
@kuv2707 @shivansh-bhatnagar18 I have mentioned the reasons for choosing custom IDs. If you feel using UUIDs is a better approach, please let me know, and I will make the desired changes accordingly.

**Comment by kuv2707** [comment]
No, this approach is fine. I just wanted you to mention the reasoning in the PR description for future reference.

**Comment by kuv2707** [comment]
No, this approach is fine. I just wanted you to mention the reasoning in the PR description for future reference.

File: backend/uno-game-engine/deck.ts [hunk 5]
```diff
@@ -57,7 +58,13 @@ export default function getShuffledCardDeck(): Array<UNOCard> {
  */
 function makeCard(type: CardType, color: CardColor, value: CardValue): UNOCard {
     //todo: Implement unique identification of cards by assigning an id to each card
```
Comment (kuv2707) [inline backend/uno-game-engine/deck.ts:60]: @sethdivyansh  Remove this todo comment.

Comment (sethdivyansh) [reply backend/uno-game-engine/deck.ts]: I have removed the comment.

**Comment by kuv2707** [comment]
Merged, thanks @sethdivyansh !

**Pull request merged**